### PR TITLE
Add __class_getitem__ to the declarative Base class

### LINKED ISSUE #7368: add __class_getitem__ to the declarative Base class
**Describe the bug**
I have a base class for SQLAlchemy tables which is shared across different projects. Therefore it uses generics.
Since the metadata, etc. is different in each project, each uses an own declarative base.
However, to have generics, the declarative base also needs to be generic.

When executing mypy, everything is fine. But during execution, it fails. As shows in the error message, a SQLAlchemy generated class is used.

**Expected behavior**
It works

**To Reproduce**

Execute `python <pyfile>`

```py
from typing import Type, TypeVar, Generic
from sqlalchemy.orm import as_declarative
from sqlalchemy import Column, Integer

T = TypeVar("T", bound="CommonBase")


class CommonBase(Generic[T]):
    @classmethod
    def boring(cls: Type[T]) -> Type[T]:
        return cls


@as_declarative()
class Base(CommonBase[T]):
    @classmethod
    def more_boring(cls: Type[T]) -> Type[T]:
        return cls


class Tab(Base["Tab"]):
    a = Column(Integer, primary_key=True)

```

**Error**

```
Traceback (most recent call last):
  File "testscript.py", line 21, in <module>
    class Tab(Base["Tab"]):
  File "/usr/lib64/python3.7/typing.py", line 251, in inner
    return func(*args, **kwds)
  File "/usr/lib64/python3.7/typing.py", line 838, in __class_getitem__
    _check_generic(cls, params)
  File "/usr/lib64/python3.7/typing.py", line 201, in _check_generic
    raise TypeError(f"{cls} is not a generic class")
TypeError: <class 'sqlalchemy.orm.decl_api.Base'> is not a generic class
```

**Versions.**
 - OS: Linux
 - Python: 3.7.1
 - SQLAlchemy: 1.4.27
 - mypy: 0.910
 - SQLAlchemy2-stubs: 0.0.2a19

**Have a nice day!**

### PR DESCRIPTION
**Pull request by kasium**
### Description
Add __class_getitem__  to the generated declarative base class

Closes #7368 

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [x] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

**Comment by kasium** [comment]
@zzzeek I'm not sure what you expect from the test. Should I just check that the code in the issue works or more, that the __class_getitem__ is generated for a declarative class? What else?

File: lib/sqlalchemy/orm/decl_api.py [hunk 1]
```diff
@@ -809,6 +809,8 @@ class Base(metaclass=DeclarativeMeta):
         class_dict["__abstract__"] = True
         if mapper:
             class_dict["__mapper_cls__"] = mapper
+        if cls.__class_getitem__:
```
Comment (zzzeek) [inline lib/sqlalchemy/orm/decl_api.py:812]: so we'd like this to port to SQLAlchemy 1.4, which still runs on Python 2, so you likely need to not assume this attribute is present.

Comment (kasium) [reply lib/sqlalchemy/orm/decl_api.py]: done

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision c935ea3524788afc212bf0bdef57b267d00972d1 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change c935ea3524788afc212bf0bdef57b267d00972d1: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3338

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 9973abfee3fb99c93c6c63f37e5bd9f99a5891b4 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset 9973abfee3fb99c93c6c63f37e5bd9f99a5891b4 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3338

File: test/orm/declarative/test_typing_py3k.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+from typing import Generic
+from typing import Type
+from typing import TypeVar
+
+from sqlalchemy import Column
+from sqlalchemy import Integer
+from sqlalchemy.orm import as_declarative
+
+
+def test_class_getitem():
```
Comment (zzzeek) [inline test/orm/declarative/test_typing_py3k.py:10]: this is good, but for it to run in our suite it has to use our base, so change this as:

from sqlalchemy.testing import fixtures

class DeclarativeBaseTest(fixtures.TestBase):
    def test_class_getiem(self):
       # ...


then we are good.   great job!

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

looks great, just needs the test reformatted so it runs

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3338

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 7db7fd869a6bb617f767fad5b71ddf7cb5f14ff5 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset 7db7fd869a6bb617f767fad5b71ddf7cb5f14ff5 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3338

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3338 has been **merged**. Congratulations! :)

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3340 has been **merged**. Congratulations! :)

**Pull request closed without merging**